### DISCUSSION TITLE
Pyright can't detect return value is not None...

### DISCUSSION THREAD

**Discussion by apocalypse2012** [microsoft/pyright#8909] [Q&A]
The following code report an error on any value returned by get_resources() that is treated like a dict.

Pyright: "values" is not a known attribute of "None"

There is no way to return a None value from that method and it is annotated to reflect that. How do keep this from occurring? I see no way for a None value to fall through...

```Python
    def get_resources(self, labels=None) -> Dict[str, Any]:
        labels = labels if labels and hasattr(labels, "__iter__") else None
        if not labels:
            return dict(self.resources)
        resource_dict = {
            label: resource_inst
            for label, resource_inst in dict(self.resources).items()
            if label in labels
        }
        return resource_dict if resource_dict else dict(self.resources)

    def is_dirty(self) -> bool:
        resource_dict = self.get_resources()
        the_resources = resource_dict.values()
        this_dirty = all(res.is_dirty() for res in the_resources)
        return this_dirty
```

**Comment by erictraut** [comment]
I presume this code is part of a class and that `resources` is an attribute with some iterable type?

Here's [my attempt](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcAsAFDnkDGANgIYDODUAYmGAFzlQ9QgCmDMAFcQVQRyg0kDGAG1UMALqUyvKABN%2BwKGn4wA%2BgKGjxDABQN%2BNYABopdAEbWGAXgByYFPwCUUALQAfJpIVPKyIPZEcEpcauq89M40zK4OycxIOkkuUHQoGlAAFox0MDAg5jkp9gBEBgZIMPwgDbV%2BLvxQnt7cCTxZUChg%2BNUMcf39AjCiKCFhltbAAHTGImKCPn1TguviBhqh%2BGkA3tuTPNWSa6b8jSiyUMC46db2Nxv3j6jzMIs2q12twYPmWTX4EAsfkG1UwczG53UAF9EbxprM%2BEDPocwpgdB99jj8J1fv8VgTNqp1FodDIDkg8HAyX4glBHOwaBNJhT6bi0lYAXpDBSoaieDAincRVA0jyicsAG50GjCQTmLbxfoSumHRkyvI0GjmYxghj0xnqp4vYxw2CSoxYswai7okBzbVm3XwchAA) to repro the issue based on the code you've provided. Pyright doesn't generate any type errors in this case, so there must be some missing details.
